cpp/src/Ice/ProxyAsync.cpp

```diff
@@ -139,6 +139,113 @@ class ProxyFlushBatchPromise : public ProxyFlushBatchAsync, public PromiseInvoke
     }
 };
 
+template<typename R>
+class InvokeOutgoingAsyncT : public OutgoingAsync
+{
+public:
+
+    using OutgoingAsync::OutgoingAsync;
+
+    void
+    invoke(const std::string& operation,
+           Ice::OperationMode mode,
+           const std::pair<const Ice::Byte*, const Ice::Byte*>& inParams,
+           const Ice::Context& context)
+    {
+        _read = [](bool ok, Ice::InputStream* stream)
+        {
+            const Ice::Byte* encaps;
+            std::int32_t sz;
+            stream->readEncapsulation(encaps, sz);
+            return R { ok, { encaps, encaps + sz } };
+        };
+
+        try
+        {
+            prepare(operation, mode, context);
+            if(inParams.first == inParams.second)
+            {
+                _os.writeEmptyEncapsulation(_encoding);
+            }
+            else
+            {
+                _os.writeEncapsulation(inParams.first, static_cast<std::int32_t>(inParams.second - inParams.first));
+            }
+            OutgoingAsync::invoke(operation);
+        }
+        catch (const std::exception&)
+        {
+            abort(std::current_exception());
+        }
+    }
+
+protected:
+
+    std::function<R(bool, Ice::InputStream*)> _read;
+};
+
+template<typename R>
+class InvokeLambdaOutgoing : public InvokeOutgoingAsyncT<R>, public LambdaInvoke
+{
+public:
+
+    InvokeLambdaOutgoing(const Ice::ObjectPrx& proxy,
+                         std::function<void(R)> response,
+                         std::function<void(std::exception_ptr)> ex,
+                         std::function<void(bool)> sent) :
+        InvokeOutgoingAsyncT<R>(proxy, false), LambdaInvoke(std::move(ex), std::move(sent))
+    {
+        if(response)
+        {
+            _response = [this, response = std::move(response)](bool ok)
+            {
+                if(this->_is.b.empty())
+                {
+                    response(R { ok, { 0, 0 }});
```

Comment:
Should we apply the 3.7 fix, or you want to further refactor this code?

https://github.com/zeroc-ice/ice/commit/49f99d1dcc042a800840e1814b6a3a5f24c1ebb4

Reply:
I plan to refactor this code much more.